File: prototype-pattern/prototype_pattern.rb
=begin
What is the Prototype Pattern?

- Creating new objects by cloning other objects
- Allows for adding of any subclass instance of a known super
  class at run time
- When there are numerous potential classes that you want to
  only use if needed at runtime
- Reduces the need for creating subclasses

@see http://www.newthinktank.com/2012/09/prototype-design-pattern-tutorial/
=end

# in ruby `clone` is available in the Kernel module

class Animal
  def make_copy
    raise NotImplementedError
  end
end

class Sheep < Animal
  def initialize
    puts 'Sheep is made'
  end

  def make_copy
    puts 'sheep is being made'

    clone
  end

  def to_s
    'Dolly is my Hero, Baaaaa'
  end
end

class CloneFactory
  def get_clone(animal_sample)
    animal_sample.make_copy
  end
end

################################################################################

animal_maker = CloneFactory.new
sally = Sheep.new
cloned_sheep = animal_maker.get_clone(sally)

puts sally
puts cloned_sheep

puts "Sally object id: #{sally.object_id}"
puts "Clone object id: #{cloned_sheep.object_id}"

File: prototype-pattern/prototype-pattern.ts
/**
 * What is the Prototype Pattern?
 *
 * - Creating new objects by cloning other objects
 * - Allows for adding of any subclass instance of a known super
 *   class at run time
 * - When there are numerous potential classes that you want to
 *   only use if needed at runtime
 * - Reduces the need for creating subclasses
 *
 * @see http://www.newthinktank.com/2012/09/prototype-design-pattern-tutorial/
 */

interface Animal {
  makeCopy(): Animal;
}

class Sheep implements Animal {
  name: string;

  public constructor() {
    console.log("Sheep is made");
  }

  makeCopy(): Animal {
    console.log("Sheep is being made");
    const sheepObject: Sheep = Object.create(this);
    return sheepObject;
  }
}

class CloneFactory {
  getClone(animalSample: Animal): Animal {
    return animalSample.makeCopy();
  }
}

//-------------------------------------------------

const animalMaker = new CloneFactory();
const sally = new Sheep();
sally.name = "sally";

const clonedSheep = <Sheep>animalMaker.getClone(sally);
clonedSheep.name = "sally clone";

console.log(sally.name);
console.log(clonedSheep.name);
